TFNetworkLayer.py

```diff
@@ -1227,6 +1227,8 @@ def custom_getter(getter, name, *args, **kwargs):
     assert not base_scope_name or base_scope_name.endswith("/")
     assert name.startswith(base_scope_name)
     rel_name = name[len(base_scope_name):]  # e.g. "rec/W" or "W"
+    if "rec/" in rel_name:
+      rel_name = name[len(base_scope_name + "rec/"):]
```

Comment:
There are multiple issues here:
* Why is it needed? There is no real reason. It should also be documented/commented properly. (And not just "because this test fails or something does not work otherwise", that does not explain anything.)
* This is too much hard-coded. Why only `"rec/"`? What about other prefixes?
* Even in the comment one line above, I esp mention that I expect it to be that way, that e.g. `"rec/W"` is totally valid here.

Reply:
Here is the error output:
```Traceback (most recent call last):
  File "tests/test_TFNetworkLayer.py", line 1080, in <module>
    line: globals()[arg]()  # assume function and execute
    locals:
      globals = <builtin> <built-in function globals>
      arg = <local> 'test_ReuseParams_rec', len = 20
  File "tests/test_TFNetworkLayer.py", line 1038, in test_ReuseParams_rec
    line: network.construct_from_dict(config.typed_dict["network"])
    locals:
      network = <local> <TFNetwork '<network via test_ReuseParams_rec>' train>
      network.construct_from_dict = <local> <bound method TFNetwork.construct_from_dict of <TFNetwork '<network via test_ReuseParams_rec>' train>>
      config = <local> <Config.Config instance at 0x7fd32078f758>
      config.typed_dict = <local> {'network': {'out1': {'loss': 'ce', 'from': ['rec_fwd'], 'class': 'softmax', 'n_out': 15}, 'out2': {'loss': 'ce', 'from': ['rec_fwd_copy'], 'class': 'softmax', 'n_out': 15}, 'rec_fwd': {'direction': 1, 'unit': 'lstmp', 'from': ['data'], 'class': 'rec', 'n_out': 300}, 'rec_fwd_copy': {'direction':..., len = 8
  File "./TFNetwork.py", line 338, in construct_from_dict
    line: self.construct_layer(net_dict, name)
    locals:
      self = <local> <TFNetwork '<network via test_ReuseParams_rec>' train>
      self.construct_layer = <local> <bound method TFNetwork.construct_layer of <TFNetwork '<network via test_ReuseParams_rec>' train>>
      net_dict = <local> {'out1': {'loss': 'ce', 'from': ['rec_fwd'], 'class': 'softmax', 'n_out': 15}, 'out2': {'loss': 'ce', 'from': ['rec_fwd_copy'], 'class': 'softmax', 'n_out': 15}, 'rec_fwd': {'direction': 1, 'unit': 'lstmp', 'from': ['data'], 'class': 'rec', 'n_out': 300}, 'rec_fwd_copy': {'direction': 1, 'from': ...
      name = <local> 'out2'
  File "./TFNetwork.py", line 407, in construct_layer
    line: layer_class.transform_config_dict(layer_desc, network=self, get_layer=get_layer)
    locals:
      layer_class = <local> <class 'TFNetworkLayer.SoftmaxLayer'>
      layer_class.transform_config_dict = <local> <bound method type.transform_config_dict of <class 'TFNetworkLayer.SoftmaxLayer'>>
      layer_desc = <local> {'loss': 'ce', 'n_out': 15}
      network = <not found>
      self = <local> <TFNetwork '<network via test_ReuseParams_rec>' train>
      get_layer = <local> <function get_layer at 0x7fd3207ae848>
  File "./TFNetworkLayer.py", line 359, in transform_config_dict
    line: d["sources"] = [
            get_layer(src_name)
            for src_name in src_names
            if not src_name == "none"]
    locals:
      d = <local> {'loss': 'ce', 'n_out': 15}
      get_layer = <local> <function get_layer at 0x7fd3207ae848>
      src_name = <local> 'rec_fwd_copy', len = 12
      src_names = <local> ['rec_fwd_copy'], _[0]: {len = 12}
  File "./TFNetwork.py", line 397, in get_layer
    line: return self.construct_layer(net_dict=net_dict, name=src_name)
    locals:
      self = <local> <TFNetwork '<network via test_ReuseParams_rec>' train>
      self.construct_layer = <local> <bound method TFNetwork.construct_layer of <TFNetwork '<network via test_ReuseParams_rec>' train>>
      net_dict = <local> {'out1': {'loss': 'ce', 'from': ['rec_fwd'], 'class': 'softmax', 'n_out': 15}, 'out2': {'loss': 'ce', 'from': ['rec_fwd_copy'], 'class': 'softmax', 'n_out': 15}, 'rec_fwd': {'direction': 1, 'unit': 'lstmp', 'from': ['data'], 'class': 'rec', 'n_out': 300}, 'rec_fwd_copy': {'direction': 1, 'from': ...
      name = <not found>
      src_name = <local> 'rec_fwd_copy', len = 12
  File "./TFNetwork.py", line 410, in construct_layer
    line: return add_layer(name=name, layer_class=layer_class, **layer_desc)
    locals:
      add_layer = <local> <bound method TFNetwork.add_layer of <TFNetwork '<network via test_ReuseParams_rec>' train>>
      name = <local> 'rec_fwd_copy', len = 12
      layer_class = <local> <class 'TFNetworkRecLayer.RecLayer'>
      layer_desc = <local> {'sources': [<SourceLayer 'data' out_type=Data(shape=(None, 100))>], 'direction': 1, 'unit': 'lstmp', 'n_out': 300, 'reuse_params': <TFNetworkLayer.ReuseParams instance at 0x7fd30c669c68>}
  File "./TFNetwork.py", line 497, in add_layer
    line: layer = self._create_layer(name=name, layer_class=layer_class, **layer_desc)
    locals:
      layer = <not found>
      self = <local> <TFNetwork '<network via test_ReuseParams_rec>' train>
      self._create_layer = <local> <bound method TFNetwork._create_layer of <TFNetwork '<network via test_ReuseParams_rec>' train>>
      name = <local> 'rec_fwd_copy', len = 12
      layer_class = <local> <class 'TFNetworkRecLayer.RecLayer'>
      layer_desc = <local> {'sources': [<SourceLayer 'data' out_type=Data(shape=(None, 100))>], 'direction': 1, 'n_out': 300, 'unit': 'lstmp', 'reuse_params': <TFNetworkLayer.ReuseParams instance at 0x7fd30c669c68>}
  File "./TFNetwork.py", line 456, in _create_layer
    line: layer = layer_class(**layer_desc)
    locals:
      layer = <not found>
      layer_class = <local> <class 'TFNetworkRecLayer.RecLayer'>
      layer_desc = <local> {'direction': 1, 'name': 'rec_fwd_copy', 'sources': [<SourceLayer 'data' out_type=Data(shape=(None, 100))>], 'unit': 'lstmp', 'network': <TFNetwork '<network via test_ReuseParams_rec>' train>, 'output': Data(name='rec_fwd_copy_output', shape=(None, 300), batch_dim_axis=1), 'n_out': 300, 'reuse_pa..., len = 8
  File "./TFNetworkRecLayer.py", line 177, in __init__
    line: y = self._get_output_native_rec_op(self.cell)
    locals:
      y = <not found>
      self = <local> <RecLayer 'rec_fwd_copy' out_type=Data(shape=(None, 300), batch_dim_axis=1)>
      self._get_output_native_rec_op = <local> <bound method RecLayer._get_output_native_rec_op of <RecLayer 'rec_fwd_copy' out_type=Data(shape=(None, 300), batch_dim_axis=1)>>
      self.cell = <local> <TFNativeOp.NativeLstmCell object at 0x7fd30c66ac50>
  File "./TFNetworkRecLayer.py", line 670, in _get_output_native_rec_op
    line: W = tf.get_variable(
            name="W", shape=(self.input_data.dim, cell.n_input_dim), dtype=tf.float32,
            initializer=self._fwd_weights_initializer)
    locals:
      W = <not found>
      tf = <global> <module 'tensorflow' from '/u/beck/programs/python/2.7.10/lib/python2.7/site-packages/tensorflow/__init__.pyc'>
      tf.get_variable = <global> <function get_variable at 0x7fd32d505c08>
      name = <not found>
      shape = <not found>
      self = <local> <RecLayer 'rec_fwd_copy' out_type=Data(shape=(None, 300), batch_dim_axis=1)>
      self.input_data = <local> Data(name='data', shape=(None, 100))
      self.input_data.dim = <local> 100
      cell = <local> <TFNativeOp.NativeLstmCell object at 0x7fd30c66ac50>
      cell.n_input_dim = <local> 1200
      dtype = <not found>
      tf.float32 = <global> tf.float32
      initializer = <not found>
      self._fwd_weights_initializer = <local> None
  File "/u/beck/programs/python/2.7.10/lib/python2.7/site-packages/tensorflow/python/ops/variable_scope.py", line 1317, in get_variable
    line: return get_variable_scope().get_variable(
              _get_default_variable_store(), name, shape=shape, dtype=dtype,
              initializer=initializer, regularizer=regularizer, trainable=trainable,
              collections=collections, caching_device=caching_device,
              partitioner=partitioner, validate_shape=validate_shape,
              use_resource=use_resource, custom_getter=custom_getter,
              constraint=constraint)
    locals:
      get_variable_scope = <global> <function get_variable_scope at 0x7fd32d505a28>
      get_variable = <global> <function get_variable at 0x7fd32d505c08>
      _get_default_variable_store = <global> <function _get_default_variable_store at 0x7fd32d505aa0>
      name = <local> 'W'
      shape = <local> (100, 1200)
      dtype = <local> tf.float32
      initializer = <local> None
      regularizer = <local> None
      trainable = <local> True
      collections = <local> None
      caching_device = <local> None
      partitioner = <local> None
      validate_shape = <local> True
      use_resource = <local> None
      custom_getter = <local> None
      constraint = <local> None
  File "/u/beck/programs/python/2.7.10/lib/python2.7/site-packages/tensorflow/python/ops/variable_scope.py", line 1079, in get_variable
    line: return var_store.get_variable(
              full_name, shape=shape, dtype=dtype, initializer=initializer,
              regularizer=regularizer, reuse=reuse, trainable=trainable,
              collections=collections, caching_device=caching_device,
              partitioner=partitioner, validate_shape=validate_shape,
              use_resource=use_resource, custom_getter=custom_getter,
              constraint=constraint)
    locals:
      var_store = <local> <tensorflow.python.ops.variable_scope._VariableStore object at 0x7fd3201b2a10>
      var_store.get_variable = <local> <bound method _VariableStore.get_variable of <tensorflow.python.ops.variable_scope._VariableStore object at 0x7fd3201b2a10>>
      full_name = <local> 'rec_fwd_copy/rec/W', len = 18
      shape = <local> (100, 1200)
      dtype = <local> tf.float32
      initializer = <local> <tensorflow.python.ops.init_ops.VarianceScaling object at 0x7fd30c66a7d0>
      regularizer = <local> None
      reuse = <local> <_ReuseMode.AUTO_REUSE: 1>
      trainable = <local> True
      collections = <local> None
      caching_device = <local> None
      partitioner = <local> None
      validate_shape = <local> True
      use_resource = <local> None
      custom_getter = <local> <function _variable_custom_getter at 0x7fd30c68a938>
      constraint = <local> None
  File "/u/beck/programs/python/2.7.10/lib/python2.7/site-packages/tensorflow/python/ops/variable_scope.py", line 417, in get_variable
    line: return custom_getter(**custom_getter_kwargs)
    locals:
      custom_getter = <local> <function _variable_custom_getter at 0x7fd30c68a938>
      custom_getter_kwargs = <local> {'validate_shape': True, 'use_resource': None, 'reuse': <_ReuseMode.AUTO_REUSE: 1>, 'regularizer': None, 'dtype': tf.float32, 'trainable': True, 'initializer': <tensorflow.python.ops.init_ops.VarianceScaling object at 0x7fd30c66a7d0>, 'shape': (100, 1200), 'getter': <function _true_getter at 0x7f..., len = 13
  File "./TFNetworkLayer.py", line 1200, in _variable_custom_getter
    line: return self.variable_custom_getter(base_layer=base_layer, **kwargs)
    locals:
      self = <local> <TFNetworkLayer.ReuseParams instance at 0x7fd30c669c68>
      self.variable_custom_getter = <local> <bound method ReuseParams.variable_custom_getter of <TFNetworkLayer.ReuseParams instance at 0x7fd30c669c68>>
      base_layer = <local> <RecLayer 'rec_fwd_copy' out_type=Data(shape=(None, 300), batch_dim_axis=1)>
      kwargs = <local> {'validate_shape': True, 'use_resource': None, 'reuse': <_ReuseMode.AUTO_REUSE: 1>, 'regularizer': None, 'dtype': tf.float32, 'trainable': True, 'initializer': <tensorflow.python.ops.init_ops.VarianceScaling object at 0x7fd30c66a7d0>, 'partitioner': None, 'shape': (100, 1200), 'getter': <function..., len = 13
  File "./TFNetworkLayer.py", line 1243, in variable_custom_getter
    line: assert rel_name in self.reuse_layer.params
    locals:
      rel_name = <local> 'rec/W'
      self = <local> <TFNetworkLayer.ReuseParams instance at 0x7fd30c669c68>
      self.reuse_layer = <local> <RecLayer 'rec_fwd' out_type=Data(shape=(None, 300), batch_dim_axis=1)>
      self.reuse_layer.params = <local> {u'b': <tf.Variable 'rec_fwd/rec/b:0' shape=(1200,) dtype=float32_ref>, u'W_re': <tf.Variable 'rec_fwd/rec/W_re:0' shape=(300, 1200) dtype=float32_ref>, u'W': <tf.Variable 'rec_fwd/rec/W:0' shape=(100, 1200) dtype=float32_ref>}
AssertionError
```

Reply:
Ok, thanks.
Still, see my comments regarding the change.
I took a look at the code myself now. I tried to fix this now via 91ff15dc408d, although I'm not 100% sure yet whether this was correct.
Maybe try yourself with that newest code. Your changes in `variable_custom_getter` became invalid in any case now, so please remove them. (But leave the pull request open, as you still have the added test code here.)